### PR TITLE
Remove gTrackGroundFlags by returning it within GameActionResult

### PR DESCRIPTION
**Pull request by Nazey**
Right now, res on `RideConstruction.cpp#L1817` does not have member GroundFlagCount, even though the new subclass `TrackPlaceActionResult` includes it. Not building because of this (obviously)

### CODE REVIEW

File: src/openrct2/actions/TrackPlaceAction.hpp [hunk 1]
```diff
@@ -68,31 +92,31 @@ DEFINE_GAME_ACTION(TrackPlaceAction, GAME_COMMAND_PLACE_TRACK, GameActionResult)
...
         {
             log_warning("Invalid direction for track placement, direction = %d", _origin.direction);
-            return std::make_unique<GameActionResult>(
-                GA_ERROR::INVALID_PARAMETERS, STR_RIDE_CONSTRUCTION_CANT_CONSTRUCT_THIS_HERE, STR_NONE);
+            return std::make_unique<TrackPlaceActionResult>(
+                GA_ERROR::INVALID_PARAMETERS, STR_NONE);
         }
 
-        auto res = std::make_unique<GameActionResult>();
+        auto res = std::make_unique<TrackPlaceActionResult>();
```
Comment (duncanspumpkin) [inline src/openrct2/actions/TrackPlaceAction.hpp]: All of the references to `gTrackGroundFlags` in this function and the Execute should be changed to our new variable `res->GroundFlagCount`

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 2]
```diff
@@ -1814,7 +1814,7 @@ static void window_ride_construction_construct(rct_window* w)
         _currentTrackSelectionFlags |= TRACK_SELECTION_FLAG_TRACK_PLACE_ACTION_QUEUED;
     }
 
-    if (gTrackGroundFlags & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND)
+    if (res->GroundFlagCount & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND)
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/RideConstruction.cpp]: Ah sorry forgot this wasn't inside the callback. You will have to do a cast to the correct type. `dynamic_cast<TrackPlaceActionResult*>(res)->GroundFlagCount` will compile.

File: data/language/en-GB.txt [hunk 3]
```diff
@@ -3585,7 +3585,7 @@ STR_6134    :Clear Scenery
 STR_6135    :Client sent invalid request
 STR_6136    :Server sent invalid request
 STR_6137    :OpenRCT2, a free and open source clone of Roller Coaster Tycoon 2.
-STR_6138    :OpenRCT2 is the work of many authors, a full list can be found in “contributors.md”. For more information, visit http://github.com/OpenRCT2/OpenRCT2
+STR_6138    :OpenRCT2 is the work of many authors, a full list can be found in “contributors.md”. For more information, visit http://github.com/OpenRCT2/OpenRCT2.
```
Comment (duncanspumpkin) [inline data/language/en-GB.txt]: Any reason for this change? This isn't really part of this PR. If you think it should be changed it would be best to address in a different PR.

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 4]
```diff
@@ -1723,10 +1723,10 @@ static void RideConstructPlacedForwardGameActionCallback(const GameAction* ga, c
     CloseConstructWindowOnCompletion(ride);
 }
 
-static void RideConstructPlacedBackwardGameActionCallback(const GameAction* ga, const GameActionResult* result)
+static void RideConstructPlacedBackwardGameActionCallback(const GameAction* ga, const TrackPlaceActionResult* result)
 {
     if (result->Error != GA_ERROR::OK)
-    {
+    {
```
Comment (duncanspumpkin) [inline src/openrct2-ui/windows/RideConstruction.cpp]: Try to avoid adding whitespace. `clang_format` will take care of this.

**Comment by duncanspumpkin** [comment]
I see that you have commited straight to develop. There is nothing wrong with that but you will tend to find it is easier to work with git if you make a feature branch off of develop and work on that. It makes it easier to rebase and fix issues. If you want to learn a bit more about it read this article. https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

When we commit this to master we will likely squash all of your 15 commits into a single one as then the history is easier to read, bisecting is easy and if required reverting is easier. Using a feature branch its easier to stop having loads of merge from develop commits.

**Comment by duncanspumpkin** [comment]
Oh and one final note add yourselves to `https://github.com/OpenRCT2/OpenRCT2/blob/develop/contributors.md` in the bug fixes section.

File: src/openrct2-ui/windows/RideConstruction.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/RideConstruction.cpp]: Ah no that wont work as this is a smart pointer.

Comment (duncanspumpkin) [reply src/openrct2-ui/windows/RideConstruction.cpp]: `dynamic_cast<TrackPlaceActionResult*>(res.get())->GroundFlagCount` now that will compile.

File: data/language/en-GB.txt (hunk 3 above)
Comment (Nazey) [reply data/language/en-GB.txt]: What I didn’t realize was that I could use the feature branch instead of directly merging. For our class, we essentially had to put our names in that string, so I just removed it. If there is a way to remove this part, I would be all for it.

**Comment by duncanspumpkin** [comment]
There is a reference to `gTrackGroundFlags` in `_legacy.cpp` for maze elements. Mazes can never be placed underground and `gTrackGroundFalgs` will always be 0 so you can safely remove the reference and choose the correct static value.

File: data/language/en-GB.txt (hunk 3 above)
Comment (duncanspumpkin) [reply data/language/en-GB.txt]: Just correct it in another commit. When we squash merge this it will be ignored.

Comment (Nazey) [reply data/language/en-GB.txt]: Ok, sweet.

**Comment by Nazey** [comment]
Ok, everything builds correctly, but the game still has trouble switching into underground mode. It will eventually switch, but it is finicky. Like, in some cases it works flawlessly, but others fails. Take this one for example. Shouldn't it change to underground mode?
Screen Shot 2019-03-20 at 7 19 54 PM

**Comment by duncanspumpkin** [comment]
So close.

Do a search for references to the variable. I know of one last place you still need to adjust. That should fix it.

**Comment by Nazey** [comment]
I swear I have looked up and down for it, but I cannot find any other references to `gTrackGroundFlags'. And I couldn't see anything with `GroundFlag` that needed to be changed either. Am I missing something in my project again? I know I am a couple 100 commits behind develop, but it shouldn't really affect the functionality of underground mode.

**Comment by duncanspumpkin** [comment]
`L118` of `src/openrct2/windows/_legacy.cpp`. You could also just delete the definition of gTrackGroundFlags and it would quickly show the problem when you compile.

**Comment by Nazey** [comment]
I didn't even realize that I hadn't added `_legacy.cpp` to my commits. That was already changed a long time ago, but it still refuses to go into underground mode.

**Comment by duncanspumpkin** [comment]
Well the mistake is still in _legacy.cpp. have a think about how this should work.

**Comment by Nazey** [comment]
Got it. Had an extra 0 that needed to be dynamically casted. That is a wicked line right there just for a variable lol.

**Comment by duncanspumpkin** [comment]
Okay this is almost finished. The last task is to delete the definition and declaration of the global variable. Then finally add yourself to the contributors file.

**Comment by Nazey** [comment]
Yeah I keep not committing the changes that I make. Need to work on that.

**Comment by duncanspumpkin** [comment]
`git commit -a` will select all changes then you won't forget.

File: src/openrct2/windows/_legacy.cpp [hunk 5]
```diff
@@ -134,7 +134,7 @@ money32 place_provisional_track_piece(
         _unkF440C5.z = z;
         _unkF440C5.direction = trackDirection;
         _currentTrackSelectionFlags |= TRACK_SELECTION_FLAG_TRACK;
-        viewport_set_visibility((gTrackGroundFlags & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND) ? 1 : 3);
+        viewport_set_visibility((0 & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND) ? 1 : 3);
```
Comment (AaronVanGeffen) [inline src/openrct2/windows/_legacy.cpp]: `((0 & TRACK_ELEMENT_LOCATION_IS_UNDERGROUND) ? 1 : 3)` will always be `3`, so this not seem correct to me.

File: src/openrct2-ui/windows/RideConstruction.cpp [hunk 6]
```diff
@@ -1790,7 +1790,7 @@ static void window_ride_construction_construct(rct_window* w)
     auto trackPlaceAction = TrackPlaceAction(
         rideIndex, trackType, { x, y, z, static_cast<uint8_t>(trackDirection) }, (properties)&0xFF, (properties >> 8) & 0x0F,
         (properties >> 12) & 0x0F, liftHillAndAlternativeState);
-
+
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/RideConstruction.cpp]:
```suggestion

```

File: src/openrct2/windows/_legacy.cpp (hunk 5 above)
Comment (Nazey) [reply src/openrct2/windows/_legacy.cpp]: Isn't this line for a maze, and mazes cannot ever be put underground?

File: src/openrct2-ui/windows/RideConstruction.cpp (hunk 4 above)
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/RideConstruction.cpp]:
```suggestion
    {
```

File: src/openrct2/actions/TrackPlaceAction.hpp [hunk 7]
```diff
@@ -19,7 +19,31 @@
...
+    TrackPlaceActionResult(GA_ERROR error, rct_string_id message)   
+    : GameActionResult(error, STR_RIDE_CONSTRUCTION_CANT_CONSTRUCT_THIS_HERE, message)
+    {
+    }
+     TrackPlaceActionResult(GA_ERROR error, rct_string_id message, uint8_t* args)  
+    : GameActionResult(error, STR_RIDE_CONSTRUCTION_CANT_CONSTRUCT_THIS_HERE, message, args)
+    {
+    }
+
+    uint8_t GroundFlag{ 0 };
```
Comment (ZehMatt) [inline src/openrct2/actions/TrackPlaceAction.hpp]: I think it would be better to call this GroundFlags

File: src/openrct2/windows/_legacy.cpp (hunk 5 above)
Comment (ZehMatt) [reply src/openrct2/windows/_legacy.cpp]: Yeah this can be always 3 as the maze does not/did not affect gTrackGroundFlags at all.

File: src/openrct2/actions/TrackPlaceAction.hpp (hunk 7 above)
Comment (Nazey) [reply src/openrct2/actions/TrackPlaceAction.hpp]: Makes more sense for consistency I guess.

**Comment by Nazey** [comment]
Anything else I should be looking for?

**Comment by janisozaur** [comment]
You still need to apply the formatting and some commits have overlong subject lines, see https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages

**Comment by AaronVanGeffen** [comment]
I think we're going to squash merge anyway due to the merge commits, so no need to adjust any commits.

**Comment by Nazey** [comment]
Yeah, from now on I'm gonna just branch off of develop and commit to that instead of develop like @duncanspumpkin said. It looks way cleaner.
